### PR TITLE
New display system and themes

### PR DESCRIPTION
**Pull request by andreasKroepelin**
This PR introduces a new system how content is displayed. Most notably, we do not rely on `set`ting fields of `page` anymore but take the display of each slide completely in our hands. This allows fine control over how each slide looks and (in my opinion) massively simplifies the creation of themes. It also means that there is now only one way to create slides, namely the `#slide` function. First and second level headings do not have a special meaning anymore.

Making use of that simplification, this PR also refactors the previous layout into a default theme. ~Visually, not much changed therefore.~ It also introduces the new theme *bipartite*.

This PR also adapts the examples and greatly extends documentation in the book.

Much inspiration came from #9 which is the intellectual ancestor to and also superseded by this PR.
I would therefore very much welcome comments by @astrale-sharp and @PgBiel who have already contributed to #9.

### CODE REVIEW

**Comment by PgBiel** [comment]
That's epic! I will try it out and leave a review over the next few days.

File: slides.typ [hunk 1]
```diff
@@ -94,74 +178,33 @@
...
-
-    show heading.where(level: 2): h => {
-        pagebreak(weak: true)
-        logical-slide.step()
-        h
-    }
-
     set text(
         size: 25pt,
     )
```
Comment (PgBiel) [inline slides.typ:190]: This should be configurable.

File: slides.typ [hunk 2]
```diff
@@ -94,74 +178,33 @@
...
-        block(
-            stroke: strokes.at(position),
-            width: 100%,
-            inset: .3em,
-            text(.5em, body)
-        )
-    }
-
     set page(
         paper: "presentation-16-9",
```
Comment (PgBiel) [inline slides.typ:193]: Configurable as well.

File: slides.typ [hunk 3]
```diff
@@ -94,74 +178,33 @@
...
-                decoration("footer")[
-                    #short-author #h(10fr)
-                    #short-title #h(1fr)
-                    #date #h(10fr)
-                    #logical-slide.display()
-                ]
-            }
-        } ),
-        footer-descent: 3em,
+        margin: 0pt,
```
Comment (PgBiel) [inline slides.typ:194]: This should be configurable too I guess, even though it is generally unwanted to do so (but power to the user)

File: slides.typ [hunk 4]
```diff
@@ -94,74 +178,33 @@
...
-                    #v(1em)
-                    #date
-                ]
-            )
-        )
-        #document_body
-    ]
+    let data = (
+        title: title,
+        author: author,
```
Comment (PgBiel) [inline slides.typ:199]: I believe we should allow passing lists of authors to themes as well (that is, lists should be the type the themes receive; if the user gives a single author as parameter to `#slides`, it should be automatically wrapped in an array).

**Review by PgBiel — CHANGES_REQUESTED** [review]
Hey-ho, here are my thoughts!

**Comment by PgBiel** [comment]
I think this will need some docs as well, but that's another step

**Comment by andreasKroepelin** [comment]
> I think this will need some docs as well, but that's another step

What kind of docs are you missing? How themes work is explained in the "create your own" section in the book. Did you think of something else still?

File: slides.typ (hunk 1 above)
Comment (andreasKroepelin) [reply slides.typ:190]: True... Also the font, which so far needs to be set extra by the user. Is that a property of the theme or of the presentation as a whole, though?

So should the user write
```typ
#show: slides.with(
    theme: theme-name(font-size: 42pt)
)
```
or
```typ
#show: slides.with(
    theme: theme-name(...),
    font-size: 42pt,
    // or possibly
    general-typography: (
        font-size: 42pt,
    )
)
```

File: slides.typ (hunk 2 above)
Comment (andreasKroepelin) [reply slides.typ:193]: Agreed, same question as with the font size though. Also would we just ask for any paper size? Or only accept the strings `"16-9"` and `°4-3"`?

File: slides.typ (hunk 3 above)
Comment (andreasKroepelin) [reply slides.typ:194]: I think this would not only be unwanted but it kinda defeats the purpose of themes, as I was imagining to give them the promise "the slide is yours, use every pixel to your delight; if it looks bad, it's your fault". So _if_ the user wanted different margins, that should be controlled via the theme. I can imagine to make the default theme accept a margin parameter.

File: slides.typ (hunk 4 above)
Comment (andreasKroepelin) [reply slides.typ:199]: Good idea. Then it's up to the theme if they want to do something fancy with it or just join the array back into one string.

File: slides.typ (hunk 1 above)
Comment (PgBiel) [reply slides.typ:190]: Well, I guess it can be sorta both - allow the theme to decide it but also let the user override it if they want

File: slides.typ (hunk 2 above)
Comment (PgBiel) [reply slides.typ:193]: Probably accept any size just in case

File: slides.typ (hunk 3 above)
Comment (PgBiel) [reply slides.typ:194]: Sure then, I agree it might be too niche. Themes probably should indeed control that

**Comment by PgBiel** [comment]
> What kind of docs are you missing?

Oh, I was just thinking of maybe adding some comments to the functions in the code
But I don't it's that urgent, just something nice to have later lol

**Comment by astrale-sharp** [comment]
Hi! Happy to see you're doing better!

Overall this looks really good !
I'm just wondering: Currently you can't overwrite the theme for just one slide, instead the only thing is you can choose a variant (if you know the string attached to it which you can't access easily from code, you have to either provide data (require looking at the source code right now) or look at the source code directly.

Instead I liked the idea of having theme be a dictionary that you can edit && inspect to make your own new theme (instead of theme variants)

that way, no string typo matters and you can style have theme variants by just defining 2 themes.

Another upside is you can base your development on the base theme (which works) edit the theme fields by fields and see with an immediate error where you got it wrong, might be useful for beginners, I wouldn't know but I'm curious.

in other words, i'd say theme is a little bit too opaque without benefits, I don't like hiding things behind string getters.

Maybe I've missed something obvious! Opinions? <3

**Comment by andreasKroepelin** [comment]
Thanks again for you feedback, @PgBiel and @astrale-sharp! I think this has come out quite nicely and will be a good foundation for further development of this template.

**Comment by astrale-sharp** [comment]
You globally convinced me! 

One last thing I'll try to add soon when I get the time is a nice error that tells you which variants are possible to choose from when you choose a bad string, it should be simple enough and is the last point that I was missing! I'll open an issue after testing a few things :angel:
